Add additional endpoints, remove non-protected endpoints

jonathanfrappier opened the pull request:
🎟️ [Jira ticket](https://hashicorp.atlassian.net/browse/SPE-178?atlOrigin=eyJpIjoiOGU5YTFlYjJkYzEwNDBjN2FkN2M1ODJhOWJmYzEwZWIiLCJwIjoiaiJ9)

🔍 [Deploy preview](https://vault-git-docs-fix-root-protected-paths-policies-hashicorp.vercel.app/vault/docs/concepts/policies#root-protected-api-endpoints)

### Description

- sys/auth/:path/tune is listed
[sys/auth/:path/tune](/vault/api-docs/system/auth#tune-auth-method)   | GET, POST, DELETE | Manage the auth methods (enable, read, delete, and tune) 
- Removed - auth/token/create-orphan is not a root-protected endpoint.
- Added missing / - There is a slash missing from sys/config/auditing/request-headers:name
- Added - sys/config-ui is not a valid endpoint. This should be sys/config/ui/headers 
- Added - sys/config/ui/headers/:name
- Removed - sys/internal/specs/openapi is not root protected!
- Removed - sys/step-down is listed as root protected but is not
- Added - sys/remount is root protected but missing from the table
- Added - sys/storage/raft/snapshot-auto/config
- Added - /sys/storage/raft/snapshot-auto/config/:name
- Spot checked several verbs, including `sys/raw` which was pointed out as an example of being incorrect, but according to our documentation is correct in the table.

Update:
- Added - sys/internal/inspect/router/ 
- Already in table specific to endpoints requiring sudo - sys/leases
- Opted to leave out since deprecated - sys/revoke-force/:prefix and sys/revoke-prefix/:prefix
- Already in table - beyond scope of this specific doc for further explanation - auth/token/create

yhyakuna commented on website/content/docs/concepts/policies.mdx:
```diff
@@ -800,32 +800,34 @@ authenticated user.
 | [sys/rotate](/vault/api-docs/system/rotate)                                                                                                            | PUT               | Trigger a rotation of the backend encryption key                                                                                                    |
 | [sys/seal](/vault/api-docs/system/seal)                                                                                                                | PUT               | Seals the Vault                                                                                                                                     |
-| [sys/step-down](/vault/api-docs/system/step-down)
+| [sys/storage/raft/snapshot-auto/config](/vault/api-docs/system/storage/raftautosnapshots#list-automated-snapshots-configs)                             | LIST              | Lists named configurations                                                                                                                          |
```
```suggestion
| [sys/step-down](/vault/api-docs/system/step-down)                             																		 | POST              | Give up a node's active status                                                                                                                      |
| [sys/storage/raft/snapshot-auto/config](/vault/api-docs/system/storage/raftautosnapshots#list-automated-snapshots-configs)                             | LIST              | Lists named configurations                                                                                                                          |
```

yhyakuna reviewed (commented):
I added `sys/step-down`.  Other than that, looks like you covered it all.

maxb commented:
> * Spot checked several verbs, including `sys/raw` which was pointed out as an example of being incorrect, but according to our documentation is correct in the table.

No, you can see at https://github.com/hashicorp/vault/blob/5cbb87e052d73c833cf0f68efe11ef6f939065f4/vault/logical_raw.go#L296 that `sys/raw` implements `ReadOperation`, `UpdateOperation`, `CreateOperation`, `DeleteOperation`, `ListOperation` - meaning that in terms of HTTP verbs it accepts GET, POST/PUT, DELETE, LIST.

But, as per what I said in #20636 

> In addition to all these issues, I'd like to flag issues with the "HTTP verb" column:
>
> Some rows mention PUT, some POST. For Vault, PUT and POST are treated identically. Therefore it is misleading to claim some APIs want one, some another.
>
> Vault list operations can be performed either via the LIST method, or via the GET method with ?list=true as a query parameter.
> 
> The mechanism in Vault for marking paths as root protected doesn't take the method into account at all, so this information isn't really needed in this table.
> 
> Some of the lists of verbs are wrong - I didn't check them all, but sys/raw definitely is, since it takes write operations which are not listed.
> 
> For all these reasons, I'd consider either dropping the HTTP verb column, or replacing it with an Operation column that categorizes operations using the ACL capabilities they require.
> 

I think you'd be best served by just removing that column entirely since it's not relevant to whether an endpoint is root-protected or not.

jonathanfrappier commented:
> > * Spot checked several verbs, including `sys/raw` which was pointed out as an example of being incorrect, but according to our documentation is correct in the table.
> 
> No, you can see at
> 
> https://github.com/hashicorp/vault/blob/5cbb87e052d73c833cf0f68efe11ef6f939065f4/vault/logical_raw.go#L296
> 
> that `sys/raw` implements `ReadOperation`, `UpdateOperation`, `CreateOperation`, `DeleteOperation`, `ListOperation` - meaning that in terms of HTTP verbs it accepts GET, POST/PUT, DELETE, LIST.
> But, as per what I said in #20650
> 
> > In addition to all these issues, I'd like to flag issues with the "HTTP verb" column:
> > Some rows mention PUT, some POST. For Vault, PUT and POST are treated identically. Therefore it is misleading to claim some APIs want one, some another.
> > Vault list operations can be performed either via the LIST method, or via the GET method with ?list=true as a query parameter.
> > The mechanism in Vault for marking paths as root protected doesn't take the method into account at all, so this information isn't really needed in this table.
> > Some of the lists of verbs are wrong - I didn't check them all, but sys/raw definitely is, since it takes write operations which are not listed.
> > For all these reasons, I'd consider either dropping the HTTP verb column, or replacing it with an Operation column that categorizes operations using the ACL capabilities they require.
> 
> I think you'd be best served by just removing that column entirely since it's not relevant to whether an endpoint is root-protected or not.

Thanks for the feedback - very much appreciated. I have addressed most of the items you originally submitted. The original request for this table we had was to include the HTTP verb. Since they are listed on the individual endpoint documentation pages, and since this page is about policy concepts someone trying to understand policies for the first time may find it helpful so I am going to leave this column.

maxb commented on website/content/docs/concepts/policies.mdx:
```diff
@@ -800,32 +800,36 @@ authenticated user.
-| [pki/root/sign-self-issued](/vault/api-docs/secret/pki#sign-self-issued)                                                                         | POST              | Use the configured CA certificate to sign a self-issued certificate ([pki secrets engine](/vault/docs/secrets/pki)) |
+| Path                                                                                                                                                   | HTTP verb         | Description                                                                                                                                         |
+| -------------------------------------------------------------------------------------------------------------------------------------------------------| ----------------- | --------------------------------------------------------------------------------------------------------------------------------------------------- |
+| [auth/token/accessors](/vault/api-docs/auth/token#list-accessors)                                                                                      | LIST              | List token accessor                                                                                                                                 |
```
The description could be a bit more useful here - how about `List token accessors for all current Vault service tokens` ?

jonathanfrappier replied:
Added to upcoming commit.

maxb commented on website/content/docs/concepts/policies.mdx:
```diff
@@ -800,32 +800,36 @@ authenticated user.
 | [sys/audit](/vault/api-docs/system/audit)                                                                                                              | GET               | List enabled audit devices                                                                                                                          |
-| [sys/audit/:path](/vault/api-docs/system/audit)                                                                                                        | PUT, DELETE       | Enable or remove an audit device                                                                                                                    |
+| [sys/audit/:path](/vault/api-docs/system/audit)                                                                                                        | POST, DELETE      | Enable or remove an audit device                                                                                                                    |
 | [sys/auth/:path](/vault/api-docs/system/auth)                                                                                                          | GET, POST, DELETE | Manage the auth methods (enable, read, delete, and tune)                                                                                            |
```
Description: This endpoint doesn't cover the 'tune' operation

maxb commented on website/content/docs/concepts/policies.mdx:
```diff
@@ -800,32 +800,36 @@ authenticated user.
-| [sys/audit/:path](/vault/api-docs/system/audit)                                                                                                        | PUT, DELETE       | Enable or remove an audit device                                                                                                                    |
+| [sys/audit/:path](/vault/api-docs/system/audit)                                                                                                        | POST, DELETE      | Enable or remove an audit device                                                                                                                    |
 | [sys/auth/:path](/vault/api-docs/system/auth)                                                                                                          | GET, POST, DELETE | Manage the auth methods (enable, read, delete, and tune)                                                                                            |
+| [sys/auth/:path/tune](/vault/api-docs/system/auth#tune-auth-method)                                                                                    | GET, POST, DELETE | Manage the auth methods (enable, read, delete, and tune)                                                                                            |
```
Verbs: No DELETE on this endpoint

Description: e.g. `Read/update auth method tuning`

jonathanfrappier replied:
Removed delete in upcoming commit.

maxb commented on website/content/docs/concepts/policies.mdx:
```diff
@@ -800,32 +800,36 @@ authenticated user.
-| [sys/internal/specs/openapi](/vault/api-docs/system/internal-specs-openapi)                                                                           | GET               | Generate an OpenAPI document of the mounted backends                                                                                                |
+| [sys/config/auditing/request-headers/:name](/vault/api-docs/system/config-auditing)                                                                    | GET, POST, DELETE | Manage the auditing headers (create, update, read and delete)                                                                                       |
+| [sys/config/cors](/vault/api-docs/system/config-cors)                                                                                                  | GET, POST, DELETE | Configure CORS setting                                                                                                                              |
+| [sys/config-ui/headers](/vault/api-docs/system/config-ui)                                                                                              | GET               | Configure the UI settings                                                                                                                           |
```
Path: Should be `sys/config/ui/headers` - slash not dash

Verb: Actually LIST

Description: e.g. `List custom HTTP headers configured to be served with the UI`

jonathanfrappier replied:
Added list and fixed slash/dash to upcoming commit

maxb commented on website/content/docs/concepts/policies.mdx:
```diff
@@ -800,32 +800,36 @@ authenticated user.
+| [sys/config/auditing/request-headers/:name](/vault/api-docs/system/config-auditing)                                                                    | GET, POST, DELETE | Manage the auditing headers (create, update, read and delete)                                                                                       |
+| [sys/config/cors](/vault/api-docs/system/config-cors)                                                                                                  | GET, POST, DELETE | Configure CORS setting                                                                                                                              |
+| [sys/config-ui/headers](/vault/api-docs/system/config-ui)                                                                                              | GET               | Configure the UI settings                                                                                                                           |
+| [sys/config-ui/headers/:name](/vault/api-docs/system/config-ui#name)                                                                                   | POST, DELETE      | Configure the values to be returned for the UI header                                                                                               |
```
Path: Should be `sys/config/ui/headers/:name` - slash not dash

Verbs: Also supports GET

Description: e.g. `Configure custom HTTP headers to be served with the UI`

jonathanfrappier replied:
Resolved in upcoming commit.

maxb commented on website/content/docs/concepts/policies.mdx:
```diff
@@ -800,32 +800,36 @@ authenticated user.
+| [sys/config/cors](/vault/api-docs/system/config-cors)                                                                                                  | GET, POST, DELETE | Configure CORS setting                                                                                                                              |
+| [sys/config-ui/headers](/vault/api-docs/system/config-ui)                                                                                              | GET               | Configure the UI settings                                                                                                                           |
+| [sys/config-ui/headers/:name](/vault/api-docs/system/config-ui#name)                                                                                   | POST, DELETE      | Configure the values to be returned for the UI header                                                                                               |
+| [sys/internal/inspect/router](/vault/api-docs/system/inspect/router  )                                                                                 | GET               | Inspect the internal components of Vault's router                                                                                                                                      |
```
Path: Should be `sys/internal/inspect/router/:tag` - it takes a path parameter

URL: I'm not sure if those trailing spaces before the closing bracket are going to upset anything... best to remove them.

maxb commented on website/content/docs/concepts/policies.mdx:
```diff
@@ -800,32 +800,36 @@ authenticated user.
+| [sys/leases/revoke-force/:prefix](/vault/api-docs/system/leases#revoke-force)                                                                          | POST              | Revoke all secrets or tokens ignoring backend errors                                                                                                |
+| [sys/leases/revoke-prefix/:prefix](/vault/api-docs/system/leases#revoke-prefix)                                                                        | POST              | Revoke all secrets generated under a given prefix                                                                                                   |
+| [sys/plugins/catalog/:type/:name](/vault/api-docs/system/plugins-catalog#register-plugin)                                                              | GET, POST, DELETE | Register a new plugin, or read/remove an existing plugin                                                                                            |
 | [sys/raw](/vault/api-docs/system/raw#list-raw)                                                                                                         | LIST, GET         | Returns a list of keys for a given path prefix                                                                                                      |
```
Verbs: Actually GET, LIST, POST, DELETE

Description: e.g. `Provides raw access to Vault's underlying storage layer`

jonathanfrappier replied:
I will open a separate ticket to review this as it only shows list and get requiring sudo at https://developer.hashicorp.com/vault/api-docs/system/raw#list-raw

maxb commented on website/content/docs/concepts/policies.mdx:
```diff
@@ -800,32 +800,36 @@ authenticated user.
+| [sys/seal](/vault/api-docs/system/seal)                                                                                                                | POST              | Seals the Vault                                                                                                                                     |
+| [sys/step-down](/vault/api-docs/system/step-down)                                                                                                      | POST              | Forces a node to give up active status                                                                                                              |
+| [sys/storage/raft/snapshot-auto/config](/vault/api-docs/system/storage/raftautosnapshots#list-automated-snapshots-configs)                             | LIST              | Lists named configurations                                                                                                                          |
+| [sys/storage/raft/snapshot-auto/config/:name](/vault/api-docs/system/storage/raftautosnapshots)                                                        | POST, GET, DELETE | Creates or updates a named configuration                                                                                                            |
```
Verbs: Truly, a minor point, but since they're ordered GET, POST, DELETE in the rest of the table rows, it would be nice to maintain the same order here.

jonathanfrappier replied:
Resolved in upcoming commit.

maxb reviewed (commented):
**The `auth/token/create` endpoint family**

The `auth/token/create` endpoint should be removed from the table. The reason is that the table is headed:

> The following paths requires a root token or sudo capability in the policy:

but this isn't true for that endpoint.

Instead, I propose that a separate paragraph be introduced below the table saying:

The `auth/token/create`, `auth/token/create-orphan` and `auth/token/create/:role_name` endpoints do not require a root token or sudo capability to use their standard functionality. However, if a root token is used, or sudo capability is granted, these endpoints unlock extra functionality:
* Allows creating tokens in child namespaces (Vault Enterprise only)
* Allows creating tokens with manually specified IDs
* Allows any policies to be attached to the new token, rather than limiting to a subset of the current token's policies
* Allows the `no_parent` option to be used to create orphan tokens via the `auth/token/create` endpoint
* Allows creating periodic tokens

**The `auth/token/revoke-orphan` endpoint**

The `auth/token/revoke-orphan` endpoint is also root protected. I missed it in my previous list, because this endpoint implements its root protection in a manual check within the code, rather than via the standard declarative mechanism, so it is not disclosed in the OpenAPI spec.

**The `sys/step-down` endpoint**

Ah, I see your commit re-adding this one, and on investigating further, I can see that this one too, is not documented as being root-protected in the OpenAPI spec, because it implements its root protection via a second unusual code flow, rather than declaratively.

**Previously mentioned still missing root-protected endpoints**

* `sys/leases`
* `sys/plugins/catalog/:name`
* `sys/plugins/catalog/:type`
* `sys/revoke-force/:prefix`
* `sys/revoke-prefix/:prefix`

maxb commented on website/content/docs/concepts/policies.mdx:
```diff
@@ -800,32 +800,36 @@ authenticated user.
 | [sys/raw](/vault/api-docs/system/raw#list-raw)                                                                                                         | LIST, GET         | Returns a list of keys for a given path prefix                                                                                                      |
-| [sys/replication/reindex](/vault/api-docs/system/replication#reindex-replication)                                                                          | POST              | Reindex the local data storage                                                                                                                      |
+| [sys/remount](/vault/api-docs/system/remount)                                                                                                          | POST              | Moves an already-mounted backend to a new mount point                                                                                               |
+| [sys/replication/reindex](/vault/api-docs/system/replication#reindex-replication)                                                                      | POST              | Reindex the local data storage                                                                                                                      |
 | [sys/replication/performance/primary/secondary-token](/vault/api-docs/system/replication/replication-performance#generate-performance-secondary-token) | POST              | Generate a performance secondary activation token                                                                                                   |
 | [sys/replication/dr/primary/secondary-token](/vault/api-docs/system/replication/replication-dr#generate-dr-secondary-token)                            | POST              | Generate a DR secondary activation token                                                                                                            |
```
I no longer work for a company with a Vault Enterprise license, but judging from the declarative definition in the OSS repository, there are more root-protected replication paths than just these three:

https://github.com/hashicorp/vault/blob/35e2c1665f009183088387532e17d02ded312e18/vault/logical_system.go#L104-L109

jonathanfrappier replied:
I will open a separate ticket to review this with engineering.

maxb commented on website/content/docs/concepts/policies.mdx:
```diff
@@ -800,32 +800,36 @@ authenticated user.
+| [sys/step-down](/vault/api-docs/system/step-down)                                                                                                      | POST              | Forces a node to give up active status                                                                                                              |
+| [sys/storage/raft/snapshot-auto/config](/vault/api-docs/system/storage/raftautosnapshots#list-automated-snapshots-configs)                             | LIST              | Lists named configurations                                                                                                                          |
+| [sys/storage/raft/snapshot-auto/config/:name](/vault/api-docs/system/storage/raftautosnapshots)                                                        | POST, GET, DELETE | Creates or updates a named configuration                                                                                                            |
 
```
Since we're keeping the HTTP verb column, adding the following could be useful:

```suggestion
~> **Note:** Vault treats the HTTP POST and PUT verbs as equivalent, so for each mention of POST in the table above, PUT may also be used. Vault uses the non-standard LIST HTTP verb, but also allows list requests to be made using the GET verb along with `?list=true` as a query parameter, so for each mention of LIST in the table above, GET with `?list=true` may also be used.
```

jonathanfrappier replied:
Added to upcoming commit.

maxb commented on website/content/docs/concepts/policies.mdx:
```diff
@@ -800,32 +800,36 @@ authenticated user.
+| [sys/config-ui/headers](/vault/api-docs/system/config-ui)                                                                                              | GET               | Configure the UI settings                                                                                                                           |
+| [sys/config-ui/headers/:name](/vault/api-docs/system/config-ui#name)                                                                                   | POST, DELETE      | Configure the values to be returned for the UI header                                                                                               |
+| [sys/internal/inspect/router](/vault/api-docs/system/inspect/router  )                                                                                 | GET               | Inspect the internal components of Vault's router                                                                                                                                      |
 | [sys/leases/lookup/:prefix](/vault/api-docs/system/leases#list-leases)                                                                                 | LIST              | List lease IDs                                                                                                                                      |
```
I see your comment:

> Already in table specific to endpoints requiring sudo - sys/leases

But I don't think that's true.

The one existing entry has a path `sys/leases/lookup/:prefix` but a link to `/vault/api-docs/system/leases#list-leases` which is a **different** endpoint!

There are actually three separate similarly named endpoints which should be in the table:

* `sys/leases` https://developer.hashicorp.com/vault/api-docs/system/leases#leases-list (GET)
* `sys/leases/lookup` https://developer.hashicorp.com/vault/api-docs/system/leases#read-lease (POST)
* `sys/leases/lookup/:prefix` https://developer.hashicorp.com/vault/api-docs/system/leases#list-leases (LIST)

All of these actually require sudo capability even though the documentation at the above links doesn't say so for all of them.

jonathanfrappier replied:
`/sys/leases/lookup/:prefix` is what is documented at https://developer.hashicorp.com/vault/api-docs/system/leases#list-leases

This endpoint returns a list of lease ids.

This endpoint requires 'sudo' capability.

Method	Path
LIST	/sys/leases/lookup/:prefix

maxb replied:
Yes, the existing documentation is quite bad at reliably identifying endpoints which require sudo, which is why I believe you should look to the code first instead.

maxb replied:
Here are the lines of code which mark these endpoints as root-protected:

https://github.com/hashicorp/vault/blob/578e62214ecabdb5e9bd6981085a8a2f71c30cd2/vault/logical_system.go#L119

https://github.com/hashicorp/vault/blob/578e62214ecabdb5e9bd6981085a8a2f71c30cd2/vault/logical_system.go#L121

jonathanfrappier replied:
Hi Max,

I will be opening a separate ticket to request help from engineering with this. I am not a developer and do not want to pretend I can interpret the source code properly.

jonathanfrappier commented:
RE auth/token/create info, opening a separate ticket to suggest reviewing and updating in the docs themselves if needed.

sys/plugins/catalog matches the docs, opening a separate ticket to request review of /vault/api-docs/system/plugins-catalog#register-plugin

jonathanfrappier commented:
Checked with engineering on open questions:

- Updated sys/raw row
- Updated sys/internal/inspect/router to specify :tag with required tags
- dr/reindex & performance/reindex are not valid. replication/reindex already listed
- sys/leases/lookup does not require sudo. sys/leases already listed
- sys/plugins/catalog/:type/:name documented as intended, leaving table to match.

jonathanfrappier commented:
I created a separate ticket to address auth/token/create. I have left it in the table currently as there are cases, according to the existing docs, that it could require sudo. New ticket shares your suggestions on additional clarity but need more time to research and collaborate with engineering.

maxb commented on website/content/docs/concepts/policies.mdx:
```diff
@@ -798,34 +798,45 @@ authenticated user.
+| [sys/leases/revoke-force/:prefix](/vault/api-docs/system/leases#revoke-force)                                                                          | POST              | Revoke all secrets or tokens ignoring backend errors                                                                                                |
+| [sys/leases/revoke-prefix/:prefix](/vault/api-docs/system/leases#revoke-prefix)                                                                        | POST              | Revoke all secrets generated under a given prefix                                                                                                   |
+| [sys/plugins/catalog/:type/:name](/vault/api-docs/system/plugins-catalog#register-plugin)                                                              | GET, POST, DELETE | Register a new plugin, or read/remove an existing plugin                                                                                            |
+| [sys/raw:path](/vault/api-docs/system/raw)                                                                                                             | GET, POST, DELETE | Used to access the raw underlying store in Vault                                                                                                    |
+| [sys/raw:prefix](/vault/api-docs/system/raw#list-raw)                                                                                                  | GET, LIST         | Returns a list keys for a given path prefix                                                                                                         |
```
This new representation of `sys/raw` does not communicate well to users.

There is a single API, which can be best described as `sys/raw/:path`, accepting GET, LIST, POST, DELETE:

```suggestion
| [sys/raw/:path](/vault/api-docs/system/raw)                                                                                                             | GET, LIST, POST, DELETE | Used to access the raw underlying store in Vault                                                                                                    |
```

It doesn't make sense to single out LIST to a separate row, use different path-parameter placeholders for the two rows, duplicate GET over both of them, or omit the slash character before the colons.

jonathanfrappier replied:
Thanks for checking in. I match the table to our docs which list `/sys/raw/:prefix` and `sys/raw:path` on https://developer.hashicorp.com/vault/api-docs/system/raw
